File: 3611-construct-the-minimum-bitwise-array-ii/construct-the-minimum-bitwise-array-ii.cpp
class Solution {
public:
    vector<int> minBitwiseArray(vector<int>& nums) {
        vector<int> ans(nums.size());
        
        for (int i = 0; i < nums.size(); i++) {
            int p = nums[i];
            
            // Case 1: Even prime (2) cannot be formed by x | (x+1)
            if (p == 2) {
                ans[i] = -1;
                continue;
            }

            // Case 2: Odd numbers
            // Find the position of the first zero bit starting from LSB
            // This tells us the range of the trailing 1s.
            int bit = 1;
            int temp = p;
            
            while ((temp & 1) == 1) {
                temp >>= 1;
                bit <<= 1;
            }
            
            // 'bit' is now just above the trailing 1s.
            // We want to subtract half of 'bit' (the MSB of the trailing 1s).
            ans[i] = p - (bit >> 1);
        }
        return ans;
    }
};

File: 3611-construct-the-minimum-bitwise-array-ii/construct-the-minimum-bitwise-array-ii.py
class Solution:
    def minBitwiseArray(self, nums: List[int]) -> List[int]:
        ans = []
        for n in nums:
            if n == 2:
                ans.append(-1)
            else:
                # Find the position of the first zero bit starting from LSB.
                # Since n is an odd prime, it has a sequence of trailing ones.
                # We iterate bits until we find a 0.
                bit = 1
                while (n & bit):
                    bit <<= 1
                
                # 'bit' is now the power of 2 representing the first zero bit.
                # We want to flip the bit immediately to its right (the MSB of trailing ones).
                # That value is 'bit' shifted right by 1.
                ans.append(n - (bit >> 1))
        return ans